components/brave_wallet/browser/internal/hd_key_ed25519_slip23.cc

```diff
@@ -0,0 +1,237 @@
+    uint32_t r = parent_prefix[i] + zr[i] + carry;
+    out[i] = r;
+    carry = r >> 8;
+  }
+
+  return out;
+}
+
+std::array<uint8_t, kSlip23ChainCodeSize> CalculateDerivedChainCode(
+    base::span<const uint8_t, crypto::hash::kSha512Size> z_hmac) {
+  std::array<uint8_t, kSlip23ChainCodeSize> chain_code;
+  base::span(chain_code)
+      .copy_from(base::span(z_hmac).last<kSlip23ChainCodeSize>());
+  return chain_code;
+}
+
+}  // namespace
+
+HDKeyEd25519Slip23::HDKeyEd25519Slip23() = default;
+HDKeyEd25519Slip23::~HDKeyEd25519Slip23() = default;
+
+// Child key derivation constructor.
+HDKeyEd25519Slip23::HDKeyEd25519Slip23(
+    base::span<const uint8_t, kSlip23ScalarSize> scalar,
+    base::span<const uint8_t, kSlip23PrefixSize> prefix,
+    base::span<const uint8_t, kSlip23ChainCodeSize> chain_code,
+    base::span<const uint8_t, kEd25519PublicKeySize> public_key) {
+  base::span(scalar_).copy_from(scalar);
+  base::span(prefix_).copy_from(prefix);
+  base::span(chain_code_).copy_from(chain_code);
+  base::span(public_key_).copy_from(public_key);
+}
+
+std::unique_ptr<HDKeyEd25519Slip23> HDKeyEd25519Slip23::DeriveChild(
+    DerivationIndex index) {
+  auto raw_index_value = index.GetValue();
+  if (!raw_index_value) {
+    return nullptr;
+  }
+
+  std::array<uint8_t, crypto::hash::kSha512Size> z_hmac = {};
+  std::array<uint8_t, crypto::hash::kSha512Size> cc_hmac = {};
+
+  if (index.is_hardened()) {
+    std::array<uint8_t, 1 + 32 + 32 + 4> data;
+    auto span_writer = base::SpanWriter(base::span(data));
+    span_writer.Skip(1u);
+    span_writer.Write(scalar_);
+    span_writer.Write(prefix_);
+    span_writer.WriteU32LittleEndian(*raw_index_value);
+
+    data[0] = 0x00;
+    z_hmac = crypto::hmac::SignSha512(chain_code_, data);
+    data[0] = 0x01;
+    cc_hmac = crypto::hmac::SignSha512(chain_code_, data);
+  } else {
+    std::array<uint8_t, 1 + 32 + 4> data;
+    auto span_writer = base::SpanWriter(base::span(data));
+    span_writer.Skip(1u);
+    span_writer.Write(public_key_);
+    span_writer.WriteU32LittleEndian(*raw_index_value);
+
+    data[0] = 0x02;
+    z_hmac = crypto::hmac::SignSha512(chain_code_, data);
+    data[0] = 0x03;
+    cc_hmac = crypto::hmac::SignSha512(chain_code_, data);
+  }
+
+  auto derived_scalar = CalculateDerivedScalar(
+      scalar_, base::span(z_hmac).first<kSlip23DerivationScalarSize>());
+
+  auto pubkey = PubkeyFromScalar(derived_scalar);
+  if (!pubkey) {
+    return nullptr;
+  }
+
+  return base::WrapUnique(new HDKeyEd25519Slip23(
+      derived_scalar,
+      CalculateDerivedPrefix(prefix_,
+                             base::span(z_hmac).last<kSlip23PrefixSize>()),
+      CalculateDerivedChainCode(cc_hmac), *pubkey));
+}
+
+// static
+std::unique_ptr<HDKeyEd25519Slip23>
+HDKeyEd25519Slip23::GenerateMasterKeyFromBip39Entropy(
+    base::span<const uint8_t> entropy) {
+  // https://github.com/cardano-foundation/CIPs/blob/master/CIP-0003/Icarus.md
```

Comment:
Is the intention to only support the Icarus derivation method? Are we going to encounter issues by not supporting the [universal master node derivation](https://github.com/satoshilabs/slips/blob/master/slip-0023.md#cardano-universal-master-node-derivation) method then?

Reply:
We support only bip39 mnemonics:
https://github.com/satoshilabs/slips/blob/master/slip-0023.md#compliance